### ISSUE TITLE
Transfer States and Queued

### ISSUE THREAD

**Issue by jackBonadies** [jpdillingham/Soulseek.NET#714]
(1) For downloads and uploads we currently have the Errored, Rejected, and TimedOut states.  Can UserOffline and CannotConnect states also be added?  They are useful and common error reasons similar to Rejected or TimedOut and would be very easy to add in the final catch clause of DownloadToStreamAsync.  I can do a pull request with these if desired.

i.e.
```
                    if (ex is UserOfflineException)
                    {
                        download.State = TransferStates.Errored | TransferStates.UserOffline;
                        throw;
                    }

                    if (ex is ConnectionException && !string.IsNullOrEmpty(ex.Message) && ex.Message.ToLower().Contains(FailedToEstablishDirectOrIndirectStringLower))
                    {
                        download.State = TransferStates.Errored | TransferStates.CannotConnect;
                    }
```

(2) When downloads are queued remotely, the calling code may desire to transition them out of this state.  For example, if we try to download from someone and are queued, they may later go offline while we are still in queue.  It might be useful to then change the transfer state from Queued to Errored  | UserOffline.  (Nicotine for example will change the text from Queued to User Logged Off.  Though note QT doesn't do this and the text will always be Queued).  I don't think there is currently a great way to do this since the only way to move out of the Queued state is to use the cancellation token and that does not allow distinction between the normal cancel or paused operation vs transition into an errored state.  

I don't know a good solution to fix this however other than adding an extra argument for Task Completion Source or similar which might be a bit messy.  Anyway, just some thoughts.

**Comment by jpdillingham** [comment]
I think for (1) a reasonable solution would be to add an `Exception` property to the download object that would contain whatever exception caused the transfer to fail.  It might still make sense to add additional state flags, I'll think about it a bit more.

(2) is a more complex issue; the expected behavior when a client goes offline is for remote users to re-enqueue any transfer that was previously queued, so if a user goes offline the associated downloads should be cancelled, then re-requested when the user comes back online.  This requires more coordination and state tracking than the library can or should provide, so implementing code should handle it.  

Retrying a failed transfer creates a new instance of the tracking object, and these are designed to be immutable outside of the library, so coordinating state changes needs to be done in some sort of wrapper; the `Transfer` object from this lib is designed to be a transient representation of the operation from the time it is initiated until it reaches any of the terminal states.

I haven't solved for this in slskd yet, so my opinion can easily change.

**Comment by jackBonadies** [comment]
Thanks for the response.  

Yes (1) would simply be for convenience, I think an Exception property would work.  

For (2) I think you are right that its not the libraries concern to coordinate these things.  Since raising this issue, I did implement what I think is a reasonable solution (without modifying the library, just in my calling code).  That is - when we are remotely queued and the peer goes offline (which is known since all relevant users are Added to get status changes), I set a flag on the TransferItem object (class I use to basically wrap the Soulseek.Net Transfer object) and then cancel the download.  Then on TransferStateChanged, I get the associated TransferItem object, and rather than replace the TransferState with that of the Transfer (which will be in state cancelled) I keep it as errored.  That way it is distinguished from a user cancelling or "pausing" the transfer, in which case on TransferStateChanged I set the state to that of the underlying Transfer object.  (The requeuing when the user goes back online is also handled in client code, though I had implemented that before opening this issue.)

All said, I believe (2) is outside the scope of the library and can be ignored.
